### DISCUSSION TITLE
How to add libraries to the javascript scripting engine?

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#14924] [Q&A] [answered]
I understand we can implement IGlobalMethodProvider, but are we able to add libraries to the scripting engine?

**Comment by Piedone** [comment]
Do you mean like full JS libraries? @sebastienros? This might be a good question for [Jint](https://github.com/sebastienros/jint), since that's what OC uses under the hood.

**Reply by weirdyang** [reply]
Yes, full libraries. I understand we can extend it using IGlobalMethodProviders, but that is C# code. I was wondering if it is possible to add full Javascript libraries.

**Comment by lampersky** [comment]
@weirdyang it is not perfect, but as a workaround you can place extarnal JS lib content before your script:

`_scriptingManager.Evaluate($"js:{externalLibraryString} {script}", null, null, GlobalMethodProviders);`

pseudcode:
![image](https://github.com/OrchardCMS/OrchardCore/assets/6403130/9961e68e-bccc-4ac0-93dc-9cf0dd87bd81)

**Comment by lampersky** [comment, marked as answer]
@weirdyang another idea is to just change a bit "createScope" method and add a way to register external modules:

![image](https://github.com/OrchardCMS/OrchardCore/assets/6403130/8a014ff2-1fb2-4001-bfe4-6b0901d9d192)

here I've just hardcoded two dummy modules, just to make sure it works:
![image](https://github.com/OrchardCMS/OrchardCore/assets/6403130/999768e6-33ee-4972-83a5-a0801e73ef98)

example:
![image](https://github.com/OrchardCMS/OrchardCore/assets/6403130/4ede462e-8626-4c25-bcc5-e9d50238308c)

**Reply by weirdyang** [reply]
thanks, will explore both! really apprecaite this
